openhands/agenthub/codeact_agent/codeact_agent.py

```diff
@@ -174,6 +182,55 @@ def reset(self) -> None:
         """Resets the CodeAct Agent."""
         super().reset()
 
+    def truncate_messages_from_middle(self, messages: list[Message], n_to_cut: int) -> list[Message]:
+        """Drop messages from the middle of the list to make it shorter."""
+        if n_to_cut <= 0:
+            return messages
+
+        truncation_message = Message(
+            role='system',
```

Comment:
Why do we do this as a system message instead of a user message like the other observations?

Reply:
Hmm good question. I think it doesn't feel the same to me as a standard observation, because it isn't really an observation returned by the environment, but more like an entity one layer above the environment saying "hey look you've been having a long conversation with the environment, and I'm gonna help you out by modifying your memories". So it's clearly not coming from the environment (who is the `user`), but it's a level above.

Reply:
Yeah that's fair enough. Overall I also don't feel that strongly. I do worry that maybe with some less performant models they might 'over attend' to system messages, but felt like we haven't noticed an issue with GPT series RE this